File: apps/tvos/Packages/CoreKit/Sources/CoreKit/PlayableChannel.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import core_api

/// The flat, `Hashable` channel currency the shell navigates and plays with. A browse `Channel`, a
/// `Favorite`-resolved channel, and a `Recent` all map to it, so one detail/play path serves every
/// entry point (home rails, drill-down, search). It carries exactly what the shell needs to
/// present, favorite/hide, and (Phase 5) play a channel — never business state, which stays in the
/// core.
public struct PlayableChannel: Hashable, Sendable, Identifiable {
  public let sourceId: Int64
  /// Stable per-source identity (favorites/hidden/recents key on this), not the churny rowid.
  public let identity: Int64
  public let name: String
  public let group: String?
  public let logo: String?
  public let locator: String

  /// Stable across a refresh, so it doubles as the SwiftUI list/focus identity.
  public var id: String { "\(sourceId)-\(identity)" }

  public init(
    sourceId: Int64, identity: Int64, name: String, group: String?, logo: String?, locator: String
  ) {
    self.sourceId = sourceId
    self.identity = identity
    self.name = name
    self.group = group
    self.logo = logo
    self.locator = locator
  }

  public init(_ channel: Channel) {
    self.init(
      sourceId: channel.sourceId,
      identity: channel.identity,
      name: channel.name,
      group: channel.groupTitle,
      logo: channel.logo,
      locator: channel.locator)
  }

  /// A recently-watched entry snapshots name + locator at play time, so it stays replayable even
  /// if the channel later left the catalog; it carries no group or logo.
  public init(_ recent: Recent) {
    self.init(
      sourceId: recent.sourceId,
      identity: recent.identity,
      name: recent.name,
      group: nil,
      logo: nil,
      locator: recent.locator)
  }
}

File: apps/androidtv/core/corekit/src/main/kotlin/dev/spidola/tv/core/corekit/PlayableChannel.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.core.corekit

import uniffi.core_api.Channel
import uniffi.core_api.MediaKind
import uniffi.core_api.Recent

/**
 * The flat channel currency the shell navigates and plays with. A browse [Channel], a
 * favorite-resolved channel, and a [Recent] all map to it, so one detail/play path serves every
 * entry point (home rails, drill-down, search). It carries exactly what the shell needs to present,
 * favorite/hide, and (Phase 5) play a channel — never business state, which stays in the core.
 */
data class PlayableChannel(
    val sourceId: Long,
    /** Stable per-source identity (favorites/hidden/recents key on this), not the churny rowid. */
    val identity: Long,
    val name: String,
    val group: String?,
    val logo: String?,
    val locator: String,
    /**
     * What the source says this is. `null` when the entry point could not say — a recent snapshots
     * no kind, so nothing downstream may claim "LIVE" without evidence (PRD §8.5).
     */
    val kind: MediaKind? = null,
) {
    companion object {
        fun of(channel: Channel): PlayableChannel =
            PlayableChannel(
                sourceId = channel.sourceId,
                identity = channel.identity,
                name = channel.name,
                group = channel.groupTitle,
                logo = channel.logo,
                locator = channel.locator,
                kind = channel.kind,
            )

        /**
         * A recently-watched entry snapshots name + locator at play time, so it stays replayable
         * even if the channel later left the catalog; it carries no group, logo, or kind.
         */
        fun of(recent: Recent): PlayableChannel =
            PlayableChannel(
                sourceId = recent.sourceId,
                identity = recent.identity,
                name = recent.name,
                group = null,
                logo = null,
                locator = recent.locator,
            )
    }
}
